File: behaviorial/memento/typescript/caretaker.ts
import { GameState } from './memento';

export class SaveHistory {
	private savedStates: GameState[] = [];

	save(state: GameState): void {
		this.savedStates.push(state);
	}

	restore(index: number): GameState | null {
		if (index >= 0 && index < this.savedStates.length) {
			return this.savedStates[index];
		}
		return null;
	}
}

File: behaviorial/memento/python/caretaker.py
class SaveHistory:
    def __init__(self):
        self.saved_states = []

    def save(self, state):
        self.saved_states.append(state)

    def restore(self, index):
        if 0 <= index < len(self.saved_states):
            return self.saved_states[index]
        return None
